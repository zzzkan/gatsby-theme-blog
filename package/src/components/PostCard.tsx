// Disable some eslint rules because Queries type is undefined in CI
/* eslint-disable @typescript-eslint/restrict-template-expressions */

import React from "react"
import { graphql, Link as GatsbyLink } from "gatsby"
import { GatsbyImage } from "gatsby-plugin-image"
import { Box, Heading, Stack, Link } from "@chakra-ui/react"
import { PostMetadata } from "./PostMetadata"
import { Tags } from "./Tags"

export const PostCard: React.FC<Queries.PostCardFragment> = ({
  slug,
  title,
  publishedDate,
  updatedDate,
  featuredImageAlt,
  timeToReadMinutes,
  publishedDateISO8601,
  updatedDateISO8601,
  featuredImage,
  tags,
}) => {
  const image = featuredImage?.childImageSharp?.gatsbyImageData
  return (
    <Box boxShadow={"md"} rounded={"xl"} overflow={"hidden"}>
      {image != null && (
        <Link
          as={GatsbyLink}
          to={slug}
          aria-label={`Move to ${slug}`}
          _hover={{ textDecoration: "none" }}
        >
          <GatsbyImage
            image={image}
            alt={featuredImageAlt ?? `Featured image for ${slug}`}
          />
        </Link>
      )}
      <Stack spacing={0} paddingX={6} paddingY={3}>
        {tags != null && <Tags tags={tags} />}
        <Link as={GatsbyLink} to={slug} aria-label={`Move to ${slug}`}>
          <Heading as={"div"} size={"md"}>
            {title}
          </Heading>
        </Link>
        <PostMetadata
          publishedDate={publishedDate}
          updatedDate={updatedDate}
          publishedDateISO8601={publishedDateISO8601}
          updatedDateISO8601={updatedDateISO8601}
          timeToReadMinutes={timeToReadMinutes}
        />
      </Stack>
    </Box>
  )
}

export const query = graphql`
  fragment PostCard on Post {
    slug
    title
    publishedDate(formatString: $dateFormatString)
    updatedDate(formatString: $dateFormatString)
    publishedDateISO8601: publishedDate(formatString: "YYYY-MM-DDTHH:mm:ss")
    updatedDateISO8601: updatedDate(formatString: "YYYY-MM-DDTHH:mm:ss")
    featuredImage {
      childImageSharp {
        gatsbyImageData(
          layout: FULL_WIDTH
          aspectRatio: $featuredImageAspectRatio
          quality: 50
        )
      }
    }
    featuredImageAlt
    tags {
      slug
      name
    }
    timeToReadMinutes
  }
`
