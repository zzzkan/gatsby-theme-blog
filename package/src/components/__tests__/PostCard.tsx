import React from "react"
import { render } from "@testing-library/react"
import { PostCard } from "../PostCard"

describe("PostCard component", () => {
  vi.mock("../PostMetadata", () => {
    return {
      PostMetadata: vi.fn().mockImplementation(() => <div>PostMetadata</div>),
    }
  })

  vi.mock("../Tags", () => {
    return {
      Tags: vi.fn().mockImplementation(() => <div>Tags</div>),
    }
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  afterAll(() => {
    vi.restoreAllMocks()
  })

  test("snapshot", () => {
    const { asFragment } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={{
          childImageSharp: { gatsbyImageData: "featuredImage" as any },
        }}
        featuredImageAlt={"featuredImageAlt"}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test("should not render image if featuredImage is null", () => {
    const { queryByRole, getAllByRole } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={null}
        featuredImageAlt={null}
      />
    )
    expect(queryByRole("img")).toBeNull()
    const links = getAllByRole("link")
    expect(links.length).toBe(1)
    links.forEach((link) => {
      expect(link).toHaveAttribute("href", "/slug")
    })
  })

  test("should not render image if featuredImage.childImageSharp is null", () => {
    const { queryByRole, getAllByRole } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={{ childImageSharp: null }}
        featuredImageAlt={null}
      />
    )
    expect(queryByRole("img")).toBeNull()
    const links = getAllByRole("link")
    expect(links.length).toBe(1)
    links.forEach((link) => {
      expect(link).toHaveAttribute("href", "/slug")
    })
  })

  test("should render image if featuredImage is not null", () => {
    const { queryByRole, getAllByRole } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={{
          childImageSharp: { gatsbyImageData: "featuredImage" as any },
        }}
        featuredImageAlt={null}
      />
    )
    expect(queryByRole("img")).toHaveAttribute("src", "featuredImage")
    const links = getAllByRole("link")
    expect(links.length).toBe(2)
    links.forEach((link) => {
      expect(link).toHaveAttribute("href", "/slug")
    })
  })

  test("should render default alt text if featuredImageAlt is null", () => {
    const { getByAltText } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={{
          childImageSharp: { gatsbyImageData: "featuredImage" as any },
        }}
        featuredImageAlt={null}
      />
    )
    expect(getByAltText("Featured image")).toBeInTheDocument()
  })

  test("should render custom alt text if featuredImageAlt is not null", () => {
    const { getByAltText } = render(
      <PostCard
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
        slug={"/slug"}
        title={"title"}
        featuredImage={{
          childImageSharp: { gatsbyImageData: "featuredImage" as any },
        }}
        featuredImageAlt={"custom featured image alt"}
      />
    )
    expect(getByAltText("custom featured image alt")).toBeInTheDocument()
  })
})
