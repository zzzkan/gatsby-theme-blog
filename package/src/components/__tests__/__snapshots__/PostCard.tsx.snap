// Vitest Snapshot v1

exports[`PostCard component > snapshot 1`] = `
<DocumentFragment>
  <article
    class="css-1luunky"
  >
    <a
      class="chakra-link css-163zzif"
      href="/slug"
    >
      <div
        class="css-1u0ptzu"
      >
        <div
          class="css-1beqgpm"
        >
          <img
            alt="featuredImageAlt"
            src="featuredImage"
          />
        </div>
      </div>
    </a>
    <div
      class="chakra-stack css-1h3nfwc"
    >
      <div>
        Tags
      </div>
      <a
        class="chakra-link css-163zzif"
        href="/slug"
      >
        <div
          class="chakra-heading css-0"
        >
          title
        </div>
      </a>
      <div>
        PostMetadata
      </div>
    </div>
  </article>
</DocumentFragment>
`;
